### PR TITLE
Introduce enable/disable experiment cohorting

### PR DESCRIPTION
**Pull request by robin-phung**
### What problem does this solve?
Disabling Cohorting: Allow existing registered participants time to convert while no longer accepting new participants into the experiment. New participants will be given the control will be given the control and will not be recorded as being apart of the experiment.

### Why is this useful?
- Gives implementers the ability to soft close experiments, allowing shifting new participants into other experiments before concluding the result of the experiment (selecting a winner).
- Allows for better accuracy of data, ~> allows participants the chance to convert if they engaged late before the end of the experiment.

### How does this solve it?
- Adds enable/disable button to the dashboard
- Provides a brief alert description describing the resulting behaviour when performed.
- Adds functions to the experiment object to check and mutate the flag stored in redis.
- Deletion of the experiment will result in the deletion of the flag (resetting it back to its default state = cohorting enabled)
- Continues to perform appropriate callbacks if the user is already an existing participant to the experiment.
- Performs no callbacks if the participant is first entering the experiment for the first time when cohorting is disabled.
- existing overrides and disables take precedence

Screen Shot 2020-04-23 at 10 20 26 PM

### CODE REVIEW

File: lib/split/trial.rb [hunk 1]
```diff
@@ -113,5 +118,9 @@ def cleanup_old_versions
     def exclude_user?
       @options[:exclude] || @experiment.start_time.nil? || @user.max_experiments_reached?(@experiment.key)
     end
+
```
Comment (slapheadted) [inline lib/split/trial.rb]: Don't need this anymore I assume?

File: spec/trial_spec.rb [hunk 2]
```diff
@@ -198,6 +198,33 @@ def expect_alternative(trial, alternative_name)
         expect(trial.alternative.name).to_not be_empty
         Split.configuration.on_trial_choose = nil
       end
+
+      it "user is assigned an alternative" do
```
Comment (slapheadted) [inline spec/trial_spec.rb]: "assigns user to an alternative"

File: spec/trial_spec.rb [hunk 3]
```diff
@@ -198,6 +198,33 @@ def expect_alternative(trial, alternative_name)
...
+      end
+
+      context "when cohorting is disabled" do
+        before(:each) { allow(experiment).to receive(:cohorting_disabled?).and_return(true) }
+
+        it "picks the control and does not run on_trial callbacks" do
+          Split.configuration.on_trial = :on_trial_callback
+
+          expect(experiment).to_not receive(:next_alternative)
+          expect(context).not_to receive(:on_trial_callback)
```
Comment (slapheadted) [inline spec/trial_spec.rb:215]: I'd probably use one for both these i.e. `not_to`.

File: spec/trial_spec.rb [hunk 4]
```diff
@@ -198,6 +198,33 @@ def expect_alternative(trial, alternative_name)
...
+          expect_alternative(trial, 'basket')
+
+          Split.configuration.enabled = true
```
Comment (slapheadted) [inline spec/trial_spec.rb:218]: I've not seen test teardown inside the `it` and after the last `expect` before. Did you consider an `after`?

File: lib/split/trial.rb (hunk 1 above)
Comment (robin-phung) [reply lib/split/trial.rb]: oops yeah 🤦

File: spec/trial_spec.rb (hunk 4 above)
Comment (robin-phung) [reply spec/trial_spec.rb:218]: I completely agree, but unfortunately I'm following the existing pattern within the spec file.

File: spec/trial_spec.rb (hunk 3 above)
Comment (robin-phung) [reply spec/trial_spec.rb:215]: as discussed following pattern existing it blocks

**Review by slapheadted — APPROVED** [review]

**Pull request merged**